lib/Sema/CSSimplify.cpp

```diff
@@ -5532,6 +5562,29 @@ ConstraintSystem::simplifyKeyPathConstraint(Type keyPathTy,
             == SolutionKind::Error)
         return SolutionKind::Error;
     }
+    if (auto contextualBGT = contextualTy->getAs<FunctionType>()) {
+      if (tryMatchRootAndValueFromFunctionType(contextualBGT) ==
+          SolutionKind::Error)
+        return SolutionKind::Error;
+    }
+  }
+
+  // If we have nothing else, and there's another constraint on our keyPathTy
+  // type variable that's some sort of conversion to a function type, use that
+  // constraint's second type.
+  if (keyPathTy->isTypeVariableOrMember()) {
+    for (auto constraint : getActiveConstraints()) {
```

Comment:
Can we delay solving this constraint until `keyPathTy` is property resolved instead of trying to guess what it might be?

Reply:
No, unfortunately. In a simple function type example like:
```
struct Struct { var I: Int }
_ = (Array<Struct>()).map(\.i)
```
The `keyPathTy` is convertible to `(Struct)->T`, but since T is unknown without figuring out the key path components, and `keyPathTy` itself is still just a type variable, the constraint system can't make progress without this code to bind the root and value types, so the result is an `error: type of expression is ambiguous without more context`.

Reply:
Interesting, this sounds like we don't model this "correctly", if there is a conversion constraint associated with keypath in theory we should be able to delay solving `keypath constraint` until type for keypath type is attempted. I'm going to pull these changes in and try it out locally to see what is up.

Reply:
I have looked at this for a bit tonight and it looks like what we need is special handling in `matchTypes` in cases where argument is a type variable representing a keypath e.g.:

```c++
@@ -2657,8 +2657,16 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     case ConstraintKind::Subtype:
     case ConstraintKind::Conversion:
     case ConstraintKind::ArgumentConversion:
-    case ConstraintKind::OperatorArgumentConversion:
+    case ConstraintKind::OperatorArgumentConversion: {
+      if (typeVar1) {
+        if (auto *locator = typeVar1->getImpl().getLocator()) {
+          if (locator->isLastElement(ConstraintLocator::KeyPathType))
+            return matchTypesBindTypeVar(typeVar1, type2, kind, flags, locator,
+                                         formUnsolvedResult);
+        }
+      }
       return formUnsolvedResult();
+    }
```

 This seems to work, although logic at the end of `simplifyKeyPathConstraint` which deals with `keyPathTy` being a function type has to be adjusted too because it doesn't have to bind function types together anymore if `keyPathTy` is already a function type.

Reply:
The logic at the end of `simplifyKeyPathConstraint` is easy to fix for function types:
```
  if (definitelyFunctionType) {
    return SolutionKind::Solved;
```
The problem then is with `KeyPath` bound generic types, which regress in a bunch of tests. This is most easily fixable if you adjust the `matchTypes` handling to only work on function types:
```
          if (locator->isLastElement(ConstraintLocator::KeyPathType) && type2->is<AnyFunctionType>())
```
That sound okay?

Reply:
It seems reasonable, sorry I didn't run all of the tests at the time. Do you happen to have an example of how they fail? I'm just curious whether we could fix that without special case in `matchTypes`.

Reply:
It looked like mostly bad key path type error diagnoses because the solver wasn't left in a state that the diagnostic code was ready for. E.g.:
```
struct S { let i: Int }
let _ : WritableKeyPath<S, Int> = \.i 
```
The key path tyvar gets bound to `WritableKeyPath` from the contextual type because of the `matchTypes` change, then the key path constraint gets simplified, sees that `i` is a `let` and tries to bind (read only) `KeyPath` as the type, and the two don't match. The regression is to `ambiguous without more context` instead of the targeted diagnosis.

Related:
```
let _: KeyPath<S, Int>? = \.i
```
Binds the key path tyvar to an optional directly, instead of resolving the path and being convertible to optional. So a failure to compile something which should work.

Reply:
I guess it would be fair to fix cases like that separately, but please leave a comment with `TODO(diagnostics):` in relevant place of `matchTypes`.